File: 5. K Most Frequent/Solution.py
import heapq
from collections import Counter 

class Solution:
    def topKFrequent(self, nums: List[int], k: int) -> List[int]:
        if k==len(nums):
            return sorted(nums,reverse=True)
        
        freq = Counter(nums)
        heap = []

        for num, count in freq.items():
            heapq.heappush(heap,(count,num))
            if len(heap)>k:
                heapq.heappop(heap)
        
        return [num for count, num in sorted(heap,key = lambda x: (-x[0],-x[1]))]


        

File: 5. K Most Frequent/Solution.js
/**
 * @param {number[]} nums
 * @param {number} k
 * @return {number[]}
 */
var topKFrequent = function(nums, k) {
    const freqMap = new Map();

    for(let num of nums){
        freqMap.set(num,(freqMap.get(num)||0)+1);
    }

    const sorted = Array.from(freqMap.entries()).sort((a,b)=>{
        if(b[1]!=a[1])return b[1]-a[1];
        return b[0]-a[0];
    });

    return sorted.slice(0,k).map(entry=>entry[0]);


};